cmd/container-disk-v2alpha/main.c

```diff
@@ -0,0 +1,157 @@
+    char *copy_path_dir;
+    char *copy_path_tmp;
+    bool health_check = false;
+
+    if (signal(SIGTERM, sig_handler) == SIG_ERR) {
+        error_log("failed to register SIGTERM callback\n");
+        exit(1);
+    }
+
+    int c;
+    while(1) {
+        static struct option long_options[] = {
+            /* These options don’t set a flag.
+                We distinguish them by their indices. */
+            {"copy-path",    required_argument, 0, 'c'},
+            {"health-check", no_argument,       0, 'p'},
+            {0, 0, 0, 0}
+        };
+        /* getopt_long stores the option index here. */
+        int option_index = 0;
+        c = getopt_long(argc, argv, "c:p", long_options, &option_index);
+
+        /* Detect the end of the options. */
+        if (c == -1) {
+            break;
+        }
+
+        switch (c) {
+            case 'c':
+                // copy_path limited by size of address.sun_path with .sock suffix
+                // copy_path + .sock + terminating null byte = 108 chars
+                // 102       + 5     + 1                     = 108
+                if (strlen(optarg) > 102) {
+                    error_log("copy path can not be longer than 102 characters\n");
+                    exit(1);
+                }
+                strncpy(copy_path, optarg, strlen(optarg));
+                copy_path_tmp = strndup(copy_path, strlen(copy_path));
+                copy_path_dir = dirname(copy_path_tmp);
+                break;
+
+            case 'p':
+                health_check = true;
+                break;
+
+            case '?':
+                exit(1);
+            default:
+                abort();
+        }
+    }
+
+    struct stat st = {0};
+    if (health_check) {
+        if (stat(READINESS_PROBE_FILE, &st) == -1) {
+            error_log("readiness probe %s does not exist, errno: %d\n", READINESS_PROBE_FILE, errno);
+            exit(1);
+        } else {
+            exit(0);
+        }
+    }
+
+    if (stat(copy_path_dir, &st) == -1) {
+        if (mkdir(copy_path_dir, 0777) != 0) {
+            error_log("failed to create disk directory %s\n", copy_path_dir);
+            exit(1);
+        }
+    }
+    free(copy_path_tmp);
+
+    struct sockaddr_un address;
+    /*
+    * For portability clear the whole structure, since some
+    * implementations have additional (nonstandard) fields in
+    * the structure.
+    */
+    memset(&address, 0, sizeof(struct sockaddr_un));
+    address.sun_family = AF_UNIX;
+    strncat(copy_path, ".sock", 5);
+    strncpy(address.sun_path, copy_path, sizeof(address.sun_path));
+
+    int fd = socket(AF_UNIX, SOCK_STREAM, 0);
+    if (fd == -1) {
+        error_log("failed to create socket on %s\n", copy_path);
+        exit(1);
+    }
+
+    if (bind(fd, (struct sockaddr*)(&address), sizeof(struct sockaddr_un)) == -1) {
+        error_log("failed to bind socket %s\n", copy_path);
+        exit(1);
+    }
+
+    if (listen(fd, LISTEN_BACKLOG) == -1) {
```

Comment:
don't we need to be accepting incoming connections? I'm surprised the net.Dial() command in isolation.go works when we don't call accept() on the server side. maybe it does somehow? 

i'm not sure what implications this might have without an accept though. For example, what happens to LISTEN_BACKLOG without an accept?

Reply:
I was actually under the impression that just `listen` is enough, because I assumed that the SYN  queue will be emptied when the client side closes the connection. Actually it sums up. So @cynepco3hahue can we please `accept` connections after `listen`?

Reply:
> Besides the code comment, I'm not completely sure of what we solve here. If memory is this inconsistent where we can't guarantee such a simple execution won't go over form 3mb to >40mb, then what hope do we have to accurately calculate the VMI overhead. This is really strange.

Right now we have an overhead of `128MB` for virt-launcher and libvirt. libvirt does not need much, so there is enough room for virt-laucher, since it seems that we did not see many reports. On the other hand reports may just not come in too often, because people don't use too many VMs with QoS guaranteed.

> 
> did we profile the process to understand how in the world this OOMs with a 40mb limit? This seems crazy. What environments are we seeing this in? is something like Transparent Hugepages enabled?

We saw it e.g. in the past on Azure, when we still had a smaller memory limit. Pushing it to 40 solved it there, but now we see it on some openshift environments (and even there only on some nodes). We should definitely not reserve more memory for something simple like this, but I also don't want to always guess for different go versions on different system. I would prefer to eliminate the problem for the containerDisk completely and maybe in the future reduce the `40M` limit.

Reply:
> So @cynepco3hahue can we please accept connections after listen

That probably involves introducing a thread again. Like you had it in the beginning :/